### PR TITLE
Issue1416

### PR DESCRIPTION
**Pull request by ptrcklv**
## Description of proposed changes
save() and load() method for Trainer to serialize the optimizer + trainer config (dependent on whether model has been fitted with that Trainer instance)
## Related issue(s)
1416
Fixes # (issue)
1416
## Test plan
test_trainer.py adapted save_load_test()
## Checklist

Need help on these? Just ask!

* [x] I have read the **CONTRIBUTING** document.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [ ] I have run `tox -e complex` and/or `tox -e spark` if appropriate.
* [x] All new and existing tests passed.

### CODE REVIEW

**Comment by henryre** [comment]
@ptrcklv thanks for taking this on! One of the maintainers will take a look soon!

File: test/classification/training/test_trainer.py [hunk 1]
```diff
@@ -216,6 +216,35 @@ def test_warmup(self):
...
+            trainer1.optimizer.state_dict()["param_groups"],
+            trainer2.optimizer.state_dict()["param_groups"],
+        )
+
+        for k in trainer1.optimizer.state_dict()["state"].keys():
+            print()
+            self.assertTrue(
+                torch.eq(
+                    trainer1.optimizer.state_dict()["state"][k]["exp_avg"],
+                    trainer2.optimizer.state_dict()["state"][k]["exp_avg"],
```
Comment (vincentschen) [inline test/classification/training/test_trainer.py]: why are we only checking equivalence of these fields? could we check the entire `state_dict`'s values?

File: test/classification/training/test_trainer.py [hunk 2]
```diff
@@ -216,6 +216,35 @@ def test_warmup(self):
         trainer.fit(model, [dataloaders[0]])
         self.assertEqual(trainer.warmup_steps, 1)
 
+    def test_save_load(self):
+        fd, checkpoint_path = tempfile.mkstemp()
```
Comment (vincentschen) [inline test/classification/training/test_trainer.py]: can we put this in a `try`/`except`/`finally` or use a context `with tempfile.NamedTemporaryFile() as f:` to ensure proper cleanup?

File: snorkel/classification/training/trainer.py [hunk 3]
```diff
@@ -509,3 +510,66 @@ def _reset_losses(self) -> None:
         """Reset the loss counters."""
         self.running_losses = defaultdict(float)
         self.running_counts = defaultdict(int)
+
+    def save(self, trainer_path: str) -> None:
+        """Save the trainer config to the specified file path in json format.
+
+        Parameters
+        ----------
+        trainer_path :
```
Comment (vincentschen) [inline snorkel/classification/training/trainer.py]: nit: no `:` here. see https://github.com/snorkel-team/snorkel/blob/master/snorkel/classification/multitask_classifier.py for example of doc strings

File: test/classification/training/test_trainer.py [hunk 4]
```diff
@@ -216,6 +216,35 @@ def test_warmup(self):
         trainer.fit(model, [dataloaders[0]])
         self.assertEqual(trainer.warmup_steps, 1)
 
+    def test_save_load(self):
```
Comment (vincentschen) [inline test/classification/training/test_trainer.py]: have you tested this with resuming training for a saved checkpoint?

File: snorkel/classification/training/trainer.py [hunk 5]
```diff
@@ -509,3 +510,66 @@ def _reset_losses(self) -> None:
...
+    ) -> None:
+        """Load trainer config and optimizer state from the specified json file path to the trainer object. The optimizer state is stored, too. However, it only makes sense if loaded with the correct model again.
+
+        Parameters
+        ----------
+        trainer_path :
+            The path to the saved trainer config to be loaded
+        model :
+            MultitaskClassifier for which the optimizer has been set. Parameters of optimizer must fit to model parameters. This model
+            shall be the model which was fit by the stored Trainer.
```
Comment (vincentschen) [inline snorkel/classification/training/trainer.py]: nit: no `:` for parameters (see above)

File: snorkel/classification/training/trainer.py [hunk 6]
```diff
@@ -509,3 +510,66 @@ def _reset_losses(self) -> None:
...
+            )
+        except BaseException:  # pragma: no cover
+            logging.warning("Saving failed... continuing anyway.")
+
+        logging.info(f"[{self.name}] Trainer config saved in {trainer_path}")
+
+    def load(
+        self, trainer_path: str, model: Optional[MultitaskClassifier] = None
+    ) -> None:
+        """Load trainer config and optimizer state from the specified json file path to the trainer object. The optimizer state is stored, too. However, it only makes sense if loaded with the correct model again.
```
Comment (vincentschen) [inline snorkel/classification/training/trainer.py]: can we add an example for expected usage, loading with with the corresponding model + resuming training? (see https://github.com/snorkel-team/snorkel/blob/master/snorkel/labeling/model/label_model.py#L366-L372 for an example, of examples (😄) in docstrings)

File: test/classification/training/test_trainer.py (hunk 2 above)
Comment (ptrcklv) [reply test/classification/training/test_trainer.py]: I put a NamedTemporaryFile, however, the tempfile.mkstemp() I copied from test_save_load() from test_multitask_classifier.py. Maybe you want to update it there, too?

File: test/classification/training/test_trainer.py (hunk 4 above)
Comment (ptrcklv) [reply test/classification/training/test_trainer.py]: yes, I have included it in the test now

File: test/classification/training/test_trainer.py (hunk 2 above)
Comment (vincentschen) [reply test/classification/training/test_trainer.py]: got it — yes, we likely need to clean up other parts of the codebase as well. :)

**Pull request merged**